sklearn/datasets/openml.py

```diff
@@ -263,6 +265,60 @@ def _convert_arff_data(arff_data, col_slice_x, col_slice_y, shape=None):
         raise ValueError('Unexpected Data Type obtained from arff.')
 
 
+def _feature_to_dtype(feature):
+    """Map feature to dtype for pandas DataFrame
+    """
+    if feature["data_type"] == "string":
+        return object
+    elif feature["data_type"] == "nominal":
+        return 'category'
+    # only numeric, integer, real are left
+    elif (feature["number_of_missing_values"] != "0" or
+          feature["data_type"] in ["numeric", "real"]):
+        return np.float64
+    elif feature["data_type"] == "integer":
+        return np.int64
+    raise ValueError("Unsupported feature: {}".format(feature))
+
+
+def _convert_arff_data_dataframe(arrf_data, all_columns, features_dict):
+    """Convert the ARFF object into a pandas DataFrame.
+
+    Parameters
+    ----------
+    arff_data : list or dict
+        as obtained from liac-arff object
+
+    all_columns : list
+        columns to return
+
+    features_dict : OrderedDict
+        map from feature to feature info from openml. This includes
+        columns that are not ignored.
+
+    Returns
+    -------
+    df : pd.DataFrame
+    """
+    pd = check_pandas_support('fetch_openml with return_frame=True')
+    df = pd.DataFrame.from_records(arrf_data['data'],
+                                   columns=list(features_dict.keys()))
+    df = df[all_columns].copy()
+
+    attributes = dict(arrf_data['attributes'])
+
+    dtypes = {}
+    for column in all_columns:
+        dtype = _feature_to_dtype(features_dict[column])
+        if dtype == object:
+            continue
+        if dtype == 'category':
+            dtype = pd.CategoricalDtype(attributes[column])
+        dtypes[column] = dtype
+
+    return df.astype(dtypes)
```

Comment:
In my original PR, I had the special case of all numeric columns (that prevents yet another copy for that specific case). 
Just mentioning, but I think it might be worth it (didn't check again with a timing though)